Feature/high dpi

arlm opened the pull request:
Full High DPI API set as described on [MSDN](https://msdn.microsoft.com/en-us/library/windows/desktop/hh447398%28v=vs.85%29.aspx).

Commit:
Adding POINT co Windows.Core

Commit:
Adding High DPI API to User32

Commit:
Adding SHCore project

Commit:
Adding High DPI API to SHCore

Commit:
Refactoring structures to Windows.ShellScalingApi project

Commit:
Adding script to create new Core Libraries

Commit:
Refactoring SHCore and User32 to reference Windows.ShellScalingApi

Commit:
General fixes for PR

Commit:
Merge branch 'develop' into feature/high-dpi

AArnott commented on src/SHCore.Desktop/SHCore.cs:
```diff
@@ -0,0 +1,103 @@
+        /// </list>
+        /// </remarks>
+        [DllImport(nameof(SHCore), SetLastError = true)]
+        public static extern HResult GetDpiForMonitor(IntPtr hmonitor, MONITOR_DPI_TYPE dpiType, out int dpiX, out int dpiY);
```
Should handles to hMonitors be a SafeMonitorHandle class? Yes, if the handle needs to be closed. No otherwise.

arlm replied:
I'll look into it and create one if needed. Wouldn't it be good to have some kind of MonitorHandle class anyway in order to differentiate the inputs? I was thinking if we needed a ProcessHandle and a WindowHandle also for other methods on Kernel32 and User32 API

AArnott replied:
The precedent has been that we only use safe handle classes where destruction of the handle is required. Yes, the type safety is nice, and perhaps we should revisit that when we're ready to make breaking changes. Feel free to file an issue to reconsider the policy on handles that we don't need to destroy.

arlm replied:
OK, I will transform that on an issue. Some handles may not have ways to Close or Destroy or it may not make sense to do that on the API, on these cases we could have something like [HandleRef ](https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.handleref%28v=vs.110%29.aspx) or a descendant of that.

MSDN Says:

> If you use platform invoke to call a managed object, and the object is not referenced elsewhere after the platform invoke call, it is possible for the garbage collector to finalize the managed object. This action releases the resource and invalidates the handle, causing the platform invoke call to fail. Wrapping a handle with HandleRef guarantees that the managed object is not garbage collected until the platform invoke call completes.
> 
> The HandleRef value type, like GCHandle, is a special type recognized by the interop marshaler. A normal, nonpinned GCHandle also prevents untimely garbage collection, yet HandleRef provides better performance. Although using HandleRef to keep an object alive for the duration of a platform invoke call is preferred, you can also use the GC.KeepAlive method for the same purpose.

AArnott replied:
>  use platform invoke to call a managed object,

I've never heard of this before. Why/how would someone use P/Invoke to call a managed object? Does that apply here?

arlm replied:
I think this sentence has mistakes. It is the other way around, at least the other places I have heard of it and how I have been using it internally. Managed things calling unmanaged things. I use it to ensure the object is bound with the reference.

Let's say I have a WinForms form. When I get the handle of the window to send to a window API I bundle a HandleRef so that the IntPtr handle is bound to the Form internal memory refrences, this way I can ensure that the handle is always present until the Form is garbage collected and the Win32 apis can safely do their thing, no GC on the IntPtr by miske, for example.

arlm replied:
Looking a little more into this issue. I found out that there is a HMONITOR type for monitor handles on Windows:

> HMONITOR:
> A handle to a display monitor.
> This type is declared in WinDef.h as follows:
> `if(WINVER >= 0x0500) typedef HANDLE HMONITOR;`

But there is no need or any way of closing it. The only Monitor handle that needs to be SafeHandle is a PhysicalMonitorSafeHandle when we do port `GetPhysicalMonitorsFromHMONITOR` and `DestroyPhysicalMonitor`

```
BOOL GetPhysicalMonitorsFromHMONITOR(
  _In_   HMONITOR hMonitor,
  _In_   DWORD dwPhysicalMonitorArraySize,
  _Out_  LPPHYSICAL_MONITOR pPhysicalMonitorArray
);

BOOL DestroyPhysicalMonitor(
  _In_  HANDLE hMonitor
);
```

Although the last function would make me think two or three times before using it (I don't want my Physical monitor destroyed), it seems just to be a poor judgement on function naming ;-) The documentation says it destroy this Physical Monitor Handle it sends you when you call the first function.

AArnott commented on src/Windows.ShellScalingApi.Desktop/ShellScallingApi+PROCESS_DPI_AWARENESS.cs:
```diff
@@ -0,0 +1,75 @@
+    /// <para>
+    /// </para>
+    /// <para>
+    /// </para>
```
Extra blank paragraphs

arlm replied:
Sorry, I'll see what I was trying to put on it and probably forgot.

AArnott commented on templates/AddNewCoreLibrary.ps1:
```diff
@@ -0,0 +1,90 @@
+<#
+    .SYNOPSIS
+    Adds the class library and NuGet projects necessary to support a new core library
```
Perhaps document what a "core library" is?

arlm replied:
Good thing also. Should I do it on the contributor.md document or on the CLI output also?

AArnott replied:
Definitely here. 
And yes, something in contributor.md sounds like it would be a good idea as I take it we expect others to need this periodically.

AArnott commented on templates/AddNewCoreLibrary.ps1:
```diff
@@ -0,0 +1,90 @@
+    Adds the class library and NuGet projects necessary to support a new core library
+    to the solution.
+
+    .PARAMETER CoreLibraryName
```
Just use "Name" as the parameter name?
If you do rename, don't forge to update the .EXAMPLE.

arlm replied:
I guess it is changed in all places it is needed because I used it to create Windows.ShellScalingApi after renaming all references of it.

By the way, I disabled on this script the part that looks for a backing DLL and generates the .txt files but I am afraid it might be used somewhere else because of the MSBuild things and/or StyleCop things. I guess I have seen references to that when I used this templating engine on another projects ([WinApi@github](https://github.com/arlm/WinApi) and [Extensions@github](https://github.com/arlm/Extensions)). Hope you don't mind I kind of "stole the idea" and ported it to these other projects. I tried to reach you on gitter before doing that but couldn't reach you.

AArnott commented on src/Windows.ShellScalingApi.Desktop/ShellScalingApi.cs:
```diff
@@ -0,0 +1,12 @@
+    /// <summary>
+    /// Exported functions from the ShellScalingApi.h header.
+    /// </summary>
+    public partial class ShellScalingApi
```
If ShellScalingApi is only for header file declarations, do we even need a test assembly?

arlm replied:
Good point. I think not. Will take it out.

AArnott replied:
Is a "core" library in your ps1 script considered a declaration-only header file library? If so, perhaps take out the unit test project from that script as well.

arlm replied:
Yes, I will. That was the idea, indeed.

AArnott commented:
Done reviewing. Thanks very much for this high quality submission. Sorry it took several days to get to it.

arlm commented on src/SHCore.Desktop/SHCore+SHELL_UI_COMPONENT.cs:
```diff
@@ -0,0 +1,29 @@
+    /// <summary>
+    /// Identifies the type of UI component that is needed in the shell
+    /// </summary>
+    public enum SHELL_UI_COMPONENT
```
Thinking over it all, I guess here I might have done a mistake. I didn't put SHELL_UI_COMPONENT under a class, so now it keeps itself hanging on the PInvoke main namespace. Should I change it so it gets under ShellScalingApi or something like that?

AArnott replied:
Good question. I guess all the types in this package are applicable to multiple libraries, right?
If we want to namespace them differently, instead of using nested types as we typically do in library packages, we should just use a namespace. That seems reasonable to me.

arlm commented:
No problem. I have sent some messages for you on gitter, not directly related to this project. And I have other PRs I'll put comments and doubts. The other PR is a little more complex than this one.

Commit:
Implementing the suggestions on the dicussion, so far

Commit:
Ohhh man, forgot to update the Solution...

vbfox commented on CONTRIBUTING.md:
```diff
@@ -35,9 +35,19 @@ in this document.
-The library should also be added to the list on the [readme](README.md).
+
+Sometimes it might be needed to add libraries to share types and structures among high-level API sets.
+When that happens you should use the `templates\AddNewLibrary.ps1` Powershell cmdlet to create a core library.
```
Isn't it `AddNewCoreLibrary.ps1` ?

arlm replied:
bad, bad CRTL+C, CTRL+V

Commit:
Fixes typo on CONTRIBUTING.md and AddNewCoreLibrary.md

Also explains a little more in details the concept and that they are not
DLL-backed.

AArnott commented on CONTRIBUTING.md:
```diff
@@ -35,9 +35,20 @@ in this document.
+Sometimes it might be needed to add libraries to share types and structures among high-level API sets.
+When that happens you should use the `templates\AddNewCoreLibrary.ps1` Powershell cmdlet to create a core library.
+
+A Core Library on PInvoke project wording is a library that only contains types and structures, no functions, classes or methods.
```
> on PInvoke project wording

change to "_in_ PInvoke project wording"

AArnott replied:
> contains types and structures

structures _are_ types. What other types did you have in mind? enums? Would it be better to say "enums and structures"?
Same thing a bit further down.

AArnott commented on CONTRIBUTING.md:
```diff
@@ -35,9 +35,20 @@ in this document.
+When that happens you should use the `templates\AddNewCoreLibrary.ps1` Powershell cmdlet to create a core library.
+
+A Core Library on PInvoke project wording is a library that only contains types and structures, no functions, classes or methods.
+This libraries are not backed by DLLs, instead they are based on C/C++ header files (.h files). They are meant to be used
```
> This libraries

"_these_ libraries"

arlm committed:
Fixing some spelling and grammar errors

arlm commented:
All the mistakes were fixed. Do you think anything else on this PR needs to be done?

AArnott commented on src/PInvoke.Win32/PInvoke.Win32.nuproj:
```diff
@@ -28,10 +28,12 @@
     <ProjectReference Include="..\Userenv.NuGet\Userenv.NuGet.nuproj" />
     <ProjectReference Include="..\UxTheme.NuGet\UxTheme.NuGet.nuproj" />
     <ProjectReference Include="..\Windows.Core.NuGet\Windows.Core.NuGet.nuproj" />
+    <ProjectReference Include="..\Windows.ShellScalingApi.NuGet\Windows.ShellScalingApi.NuGet.nuproj" />
```
nit: This reference can be removed. It will be a transitive dependency anyway.

AArnott commented:
Looks good. Thanks!
